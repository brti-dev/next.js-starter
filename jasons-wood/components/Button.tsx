import React from 'react'
import Link from 'next/link'

export type ButtonProps = {
    type?: 'button' | 'reset' | 'submit'
    variant?: 'text' | 'contained' | 'outlined' | 'link' | 'icon' | 'close'
    color?:
        | 'default'
        | 'primary'
        | 'secondary'
        | 'red'
        | 'green'
        | 'dark'
        | 'light'
    size?: 'small' | 'medium' | 'large'
    className?: string
    to?: string
    disabled?: boolean
    loading?: boolean
    children: React.ReactNode
} & React.ComponentPropsWithRef<'button'> &
    React.ComponentPropsWithRef<'a'>

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
    (props, ref) => {
        const {
            type = 'button',
            variant = 'text',
            color = 'default',
            size = 'medium',
            className,
            to,
            disabled = false,
            loading = false,
            children,
            ...rest
        } = props

        const classNames = [
            'button',
            `variant-${variant}`,
            `color-${color}`,
            `size-${size}`,
            className && className,
        ]

        if (to) {
            return (
                <Link href={to}>
                    <a className={classNames.join(' ')} {...rest}>
                        {children}
                    </a>
                </Link>
            )
        }

        return (
            <button
                type={type}
                className={classNames.join(' ')}
                disabled={disabled || loading ? true : undefined}
                data-loading={loading ? true : undefined}
                ref={ref}
                {...rest}
            >
                {children}
            </button>
        )
    }
)

export default Button
