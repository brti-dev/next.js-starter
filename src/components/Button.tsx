import React from 'react'
import Link from 'next/link'

export type ButtonProps = {
  type?: 'button' | 'reset' | 'submit'
  variant?: 'text' | 'contained' | 'outlined' | 'link' | 'close'
  color?:
    | 'default'
    | 'primary'
    | 'secondary'
    | 'red'
    | 'green'
    | 'dark'
    | 'light'
  size?: 'small' | 'medium' | 'large'
  className?: string
  to?: string
  disabled?: boolean
  loading?: boolean
  icon?: boolean
  children: React.ReactNode
} & React.ComponentPropsWithRef<'button'> &
  React.ComponentPropsWithRef<'a'>

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      type = 'button',
      variant = 'text',
      color = 'default',
      size = 'medium',
      className,
      to,
      disabled = false,
      loading = false,
      icon = false,
      children,
      ...rest
    } = props

    const classNames = [
      'button',
      `variant-${variant}`,
      `color-${color}`,
      `size-${size}`,
      icon && 'icon',
      className && className,
    ]

    if (to) {
      return (
        <Link href={to}>
          <a className={classNames.join(' ')} {...rest}>
            {children}
          </a>
        </Link>
      )
    }

    return (
      <button
        type={type}
        className={classNames.join(' ')}
        disabled={disabled || loading ? true : undefined}
        data-loading={loading ? true : undefined}
        ref={ref}
        {...rest}
      >
        {children}
      </button>
    )
  }
)

export default Button
