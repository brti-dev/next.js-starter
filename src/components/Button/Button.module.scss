@use 'styles/theme' as *;

.button {
  display: inline-flex;
  gap: 0.5em;
  vertical-align: middle;
  align-items: center;
  padding: 0 0.95rem;
  cursor: pointer;
  background: transparent;

  &[data-loading='true'] {
    position: relative;
    overflow: hidden;
    background: var(--color-surface);
    cursor: wait;

    &::after {
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      content: '';
      position: absolute;
      animation: wave 1.6s linear 0.5s infinite;
      transform: translateX(-100%);
      background: linear-gradient(
        90deg,
        transparent,
        rgba(255, 255, 255, 0.15),
        transparent
      );
    }
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.75;
  }
}

.small {
  font-size: 80%;
}

.large {
  font-size: 120%;
}

.text,
.contained,
.outlined,
.link {
  &:hover {
    background: var(--color-surface);
  }
}

.text {
  &:hover:not(:focus) {
    box-shadow: none;
  }
}

.icon {
  padding: 8px;
  box-shadow: none !important;
  border-radius: 50%;
}

.contained {
  &[class~='color-default'] {
    background: var(--color-surface);
    &:hover {
      background: var(--color-surface-hover);
    }
  }

  &[class~='color-primary'] {
    background: var(--color-primary);
    color: white;
    &:hover {
      background: var(--color-primary-hover);
    }
  }

  &[class~='color-secondary'] {
    background: var(--color-secondary);
    color: white;
    &:hover {
      background: var(--color-secondary-hover);
    }
  }

  &[class~='color-red'] {
    background: $red;
    color: white;
  }

  &[class~='color-green'] {
    background: $green;
    color: white;
  }
}

.outlined {
  line-height: calc(var(--line-height) - 2px);
  border: 1px solid var(--color-comment);
}

@keyframes wave {
  0% {
    transform: translateX(-100%);
  }
  60% {
    transform: translateX(100%);
  }
  100% {
    transform: translateX(100%);
  }
}
