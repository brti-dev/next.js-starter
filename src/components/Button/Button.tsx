import { forwardRef } from 'react'
import Link from 'next/link'

import { Color, Variant } from 'interfaces/theme'
import classnames from 'lib/classnames'

type Percent = `${number}%`

export type ButtonProps = {
  type?: 'button' | 'reset' | 'submit'
  variant?: Variant | 'link' | 'close'
  color?: Color
  size?: 'small' | 'medium' | 'large'
  width?: number | Percent
  className?: string
  to?: string
  disabled?: boolean
  loading?: boolean
  icon?: boolean
  children: React.ReactNode
} & React.ComponentPropsWithRef<'button'> &
  React.ComponentPropsWithRef<'a'>

const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    type = 'button',
    variant = 'default',
    color = 'default',
    size = 'medium',
    width,
    className,
    to,
    disabled = false,
    loading = false,
    icon = false,
    children,
    ...rest
  } = props

  const style: React.CSSProperties = {}
  if (typeof width === 'number') {
    style.width = `${width}px`
  } else if (typeof width === 'string') {
    // Percent
    style.width = width
  }

  const classNameString = classnames(
    'button', // Give access to global button style shared with other inputs
    `variant--${variant}`,
    `color--${color}`,
    `size--${size}`,
    icon && 'button__icon',
    className
  )

  if (to) {
    return (
      <Link href={to}>
        <a className={classNameString} {...rest}>
          <span className="text-content">{children}</span>
        </a>
      </Link>
    )
  }

  return (
    <button
      type={type}
      className={classNameString}
      disabled={disabled || loading ? true : undefined}
      data-loading={loading ? 'true' : undefined}
      style={style}
      ref={ref}
      {...rest}
    >
      <span className={`${icon ? 'icon' : 'text'}-content`}>{children}</span>
    </button>
  )
})

export default Button
