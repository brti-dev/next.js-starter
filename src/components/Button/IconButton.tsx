import Tooltip from '../Tooltip'
import Button, { ButtonProps } from '.'

export type IconButtonProps = ButtonProps & {
  tooltip?: string
}

function IconButton({ children, ...rest }: any) {
  return (
    <Button icon={true} {...rest}>
      {children}
    </Button>
  )
}

function CustomIconButton({ tooltip, ...rest }: IconButtonProps) {
  return (
    <>
      {tooltip ? (
        <Tooltip label={tooltip}>
          <IconButton aria-label={tooltip} {...rest} />
        </Tooltip>
      ) : (
        <IconButton {...rest} />
      )}
    </>
  )
}
export default CustomIconButton
