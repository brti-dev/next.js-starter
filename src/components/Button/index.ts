import Button, { ButtonProps } from './Button'
import IconButton, { IconButtonProps } from './IconButton'

export { Button as default, IconButton }
export type { ButtonProps, IconButtonProps }
