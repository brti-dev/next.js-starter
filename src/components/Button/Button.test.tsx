import '@testing-library/jest-dom'

import { render, screen } from '../../../test-utils'
import Button from './Button'

test('shows spinner & disabled if loading', () => {
  const { getByText } = render(<Button loading>Loading</Button>)

  const expBtn = expect(getByText('Loading'))
  expBtn.toHaveAttribute('data-loading')
  expBtn.toHaveAttribute('disabled')
})

test('has the proper aria attributes', () => {
  const { rerender } = render(<Button>button</Button>)

  // button has role="button"
  let button = screen.getByRole('button')
  expect(button).not.toHaveAttribute('aria-disabled')

  // loading sets aria-disabled to true
  rerender(<Button loading>button</Button>)
  button = screen.getByRole('button')
  expect(button).toHaveAttribute('disabled')
})

test('has the proper `type` attribute', () => {
  const { getByRole, rerender } = render(<Button>button</Button>)

  expect(getByRole('button')).toHaveAttribute('type', 'button')

  rerender(
    <Button data-testid="btn" type="submit">
      button
    </Button>
  )

  expect(getByRole('button')).toHaveAttribute('type', 'submit')

  rerender(
    <Button data-testid="btn" type="reset">
      button
    </Button>
  )

  expect(getByRole('button')).toHaveAttribute('type', 'reset')
})

test('should be disabled', () => {
  const { getByRole } = render(<Button disabled>button</Button>)
  expect(getByRole('button')).toHaveAttribute('disabled')
})
