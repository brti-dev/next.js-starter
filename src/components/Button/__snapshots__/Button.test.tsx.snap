// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<button
  className="button variant--contained color--primary size--small"
  style={
    Object {
      "width": "50px",
    }
  }
  type="button"
>
  <span
    className="text-content"
  >
    Button
  </span>
</button>
`;

exports[`should render icon button correctly 1`] = `
<button
  className="button variant--default color--default size--medium button__icon"
  style={Object {}}
  type="button"
>
  <span
    className="icon-content"
  >
    icon
  </span>
</button>
`;

exports[`should work with different colors 1`] = `
Array [
  <button
    className="button variant--default color--primary size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      primary
    </span>
  </button>,
  <button
    className="button variant--default color--secondary size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      secondary
    </span>
  </button>,
  <button
    className="button variant--default color--error size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      error
    </span>
  </button>,
  <button
    className="button variant--default color--warning size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      warning
    </span>
  </button>,
  <button
    className="button variant--default color--info size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      info
    </span>
  </button>,
  <button
    className="button variant--default color--success size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      success
    </span>
  </button>,
  <button
    className="button variant--default color--dark size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      dark
    </span>
  </button>,
  <button
    className="button variant--default color--light size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      light
    </span>
  </button>,
]
`;

exports[`should work with different sizes 1`] = `
Array [
  <button
    className="button variant--default color--default size--small"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      small
    </span>
  </button>,
  <button
    className="button variant--default color--default size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      medium
    </span>
  </button>,
  <button
    className="button variant--default color--default size--large"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      large
    </span>
  </button>,
]
`;

exports[`should work with different variants 1`] = `
Array [
  <button
    className="button variant--default color--default size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      default
    </span>
  </button>,
  <button
    className="button variant--outlined color--default size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      outlined
    </span>
  </button>,
  <button
    className="button variant--contained color--default size--medium"
    style={Object {}}
    type="button"
  >
    <span
      className="text-content"
    >
      contained
    </span>
  </button>,
]
`;
